### PR TITLE
refactor: split timed messages feature into a separate handler

### PR DESCRIPTION
**Pull request by atomrc**
:warning: important architectural changes in there.

The idea in this PR is to split feature related code to specific classes (called `handler`).

The few benefits of this approach is:
- lesser code in the `ConversationRepository`;
- small and easy testable handler classes;
- Possibility to easily toggle on/off a feature;
- cleaner separation of concerns (the `Repo` is the proxy to the external world and the database keeper while the handlers are the feature specific implementation on the entity level);

### CODE REVIEW

File: app/script/conversation/ConversationStateHandler.js [hunk 1]
```diff
@@ -114,43 +121,16 @@ z.conversation.ConversationStateHandler = class ConversationStateHandler {
...
-      return Promise.reject(new Error('Conversation Repository Event Handling: Event missing'));
-    }
+  _mapConversationAccessState(conversationEntity, eventJson) {
+    this.conversationMapper.mapAccessState(conversationEntity, eventJson.access, eventJson.access_role);
+  }
 
-    const {conversation: conversationId, data: eventData, type} = eventJson;
-
-    return this.conversationRepository
-      .get_conversation_by_id(conversationId)
```
Comment (atomrc) [inline app/script/conversation/ConversationStateHandler.js:132]: Fetching the conv in the DB is now refactored in the Repo

File: app/script/conversation/AbstractConversationEventHandler.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  /**
+   * Abtract class that represents an entity that can
+   * react to a conversation event
+   */
+  constructor() {
+    this.eventHandlingConfig = {};
+  }
+
+  /**
+   * adds an eventhandling config to the current instance
```
Comment (gregor) [inline app/script/conversation/AbstractConversationEventHandler.js]: adds -> Adds and full stop at the end

File: app/script/conversation/AbstractConversationEventHandler.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * @typedef {Object} EventHandlingConfig - Object representing conversation event handlers
+ * @property {() => void} [eventId: string] - an event handler function
+ */
+
+z.conversation.AbstractConversationEventHandler = class AbstractConversationEventHandler {
+  /**
+   * Abtract class that represents an entity that can
+   * react to a conversation event
```
Comment (gregor) [inline app/script/conversation/AbstractConversationEventHandler.js]: Full stop

File: app/script/conversation/AbstractConversationEventHandler.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+   * adds an eventhandling config to the current instance
+   * @param {EventHandlingConfig} eventHandlingConfig - Config containing events name and the associated callback
+   * @returns {void}
+   */
+  setEventHandlingConfig(eventHandlingConfig) {
+    this.eventHandlingConfig = eventHandlingConfig;
+  }
+
+  /**
+   * Handles a conversation event
```
Comment (gregor) [inline app/script/conversation/AbstractConversationEventHandler.js]: Full stop

File: app/script/conversation/ConversationEphemeralHandler.js [hunk 5]
```diff
@@ -22,17 +22,22 @@
...
-  constructor(conversationService, onMessageTimeout) {
+  constructor(conversationService, conversationMapper, onMessageTimeout) {
+    super();
+    this.setEventHandlingConfig({
+      [z.event.Backend.CONVERSATION.MESSAGE_TIMER_UPDATE]: this._updateEphemeralTimer.bind(this),
+    });
     this._updateTimedMessages = this._updateTimedMessages.bind(this);
-
     this.conversationService = conversationService;
+    this.conversationMapper = conversationMapper;
```
Comment (gregor) [inline app/script/conversation/ConversationEphemeralHandler.js]: ABC

File: app/script/conversation/ConversationEphemeralHandler.js [hunk 6]
```diff
@@ -207,4 +212,18 @@ z.conversation.ConversationEphemeralHandler = class ConversationEphemeralHandler
     });
     this.logger.info(`Obfuscated text message '${messageEntity.id}'`);
   }
+
+  /**
+   * Updates the ephemeral timer of a conversation when an timer-update message is received
```
Comment (gregor) [inline app/script/conversation/ConversationEphemeralHandler.js]: Full stop

File: app/script/conversation/ConversationRepository.js [hunk 7]
```diff
@@ -2735,7 +2748,57 @@ z.conversation.ConversationRepository = class ConversationRepository {
       });
   }
 
-  _handledConversationEvent(entityObject = {}, eventSource, previouslyArchived) {
+  /**
+   * calls the feature specific event handler on the current event being handled
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.js]: calls -> Calls and full stop

File: app/script/conversation/ConversationRepository.js [hunk 8]
```diff
@@ -2735,7 +2748,57 @@ z.conversation.ConversationRepository = class ConversationRepository {
...
+  /**
+   * Add relevant events to the conversation entity in order for it to be displayed on the UI.
+   * Does nothing if the event type doesn't need to be persisted in memory
+   *
+   * @private
+   * @param {Conversation} conversationEntity - Conversation to add the event to
+   * @param {Object} eventJson - JSON data of the event
+   * @returns {Promise} Resolves when event was handled
+   */
+  _handleEventPersistence(conversationEntity, eventJson) {
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.js]: Let's rename this. Persistence is not really a good choice as we aren't storing it in the database here.

File: app/script/conversation/ConversationRepository.js [hunk 9]
```diff
@@ -2690,41 +2684,60 @@ z.conversation.ConversationRepository = class ConversationRepository {
...
+            break;
           default:
-            return this._onAddEvent(conversationEntity, eventJson);
+            handledPromise = Promise.resolve();
         }
+        return handledPromise.then(() => conversationEntity);
       })
-      .then((entityObject = {}) => this._handledConversationEvent(entityObject, eventSource, previouslyArchived))
+
+      .then(conversationEntity => this._handleEventPersistence(conversationEntity, eventJson))
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.js]: This kind of takes over the default case for the switch case. It should not. Let's add explicit handling for the message types that previously fell into the default case.  Call message, verification states changes would all be missing in the UI at this point. Let's move the `_addEventToConversation` back to the individual conversation event handlers and use the `_onAddEvent` not as the default case but explicitly for the expected event types.

File: app/script/conversation/ConversationRepository.js [hunk 10]
```diff
@@ -2733,7 +2746,57 @@ z.conversation.ConversationRepository = class ConversationRepository {
...
+      return Promise.resolve({});
+    }
+    return this._addEventToConversation(conversationEntity, eventJson).then(messageEntity => {
+      return {conversationEntity, messageEntity};
+    });
+  }
+
+  /**
+   * Update a few conversation status depending on the event that was handled
+   * also handles notifications if needed
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.js]: Full stop

File: app/script/conversation/ConversationRepository.js [hunk 11]
```diff
@@ -2690,41 +2684,12 @@ z.conversation.ConversationRepository = class ConversationRepository {
...
-            return this._on_reaction(conversationEntity, eventJson);
-          default:
-            return this._onAddEvent(conversationEntity, eventJson);
-        }
+        return conversationEntity;
       })
-      .then((entityObject = {}) => this._handledConversationEvent(entityObject, eventSource, previouslyArchived))
+      .then(conversationEntity => this._triggerFeatureEventHandlers(conversationEntity, eventJson, eventSource))
+      .then(() => this._reactToConverationEvent(conversationEntity, eventJson))
+
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.js]: Remove empty line

File: app/script/conversation/ConversationRepository.js [hunk 12]
```diff
@@ -2733,7 +2698,88 @@ z.conversation.ConversationRepository = class ConversationRepository {
...
+        return this._on_message_add(conversationEntity, eventJson);
+      case z.event.Client.CONVERSATION.MESSAGE_DELETE:
+        return this._onMessageDeleted(conversationEntity, eventJson);
+      case z.event.Client.CONVERSATION.MESSAGE_HIDDEN:
+        return this._onMessageHidden(eventJson);
+      case z.event.Client.CONVERSATION.ONE2ONE_CREATION:
+        return this._on1to1Creation(conversationEntity, eventJson);
+      case z.event.Client.CONVERSATION.REACTION:
+        return this._on_reaction(conversationEntity, eventJson);
+      case z.event.Client.CONVERSATION.ASSET_ADD:
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.js]: We already have that in the switch case

File: app/script/conversation/ConversationRepository.js [hunk 13]
```diff
@@ -2733,7 +2698,88 @@ z.conversation.ConversationRepository = class ConversationRepository {
...
+      case z.event.Client.CONVERSATION.LOCATION:
+      case z.event.Client.CONVERSATION.MISSED_MESSAGES:
+      case z.event.Client.CONVERSATION.UNABLE_TO_DECRYPT:
+      case z.event.Client.CONVERSATION.VERIFICATION:
+      case z.event.Backend.CONVERSATION.CONNECT_REQUEST:
+      case z.event.Backend.CONVERSATION.DELETE:
+      case z.event.Backend.CONVERSATION.MEMBER_JOIN:
+      case z.event.Backend.CONVERSATION.MESSAGE_TIMER_UPDATE:
+      case z.event.Backend.CONVERSATION.OTR_MESSAGE_ADD:
+      case z.event.Backend.CONVERSATION.TYPING:
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.js]: We need to clean this list.

**Pull request merged**